pkg/gpu/nvml/devices.go

```diff
@@ -0,0 +1,174 @@
+	}
+
+	cores, ret := dev.GetNumGpuCores()
+	if ret != nvml.SUCCESS {
+		return nil, fmt.Errorf("error getting core count: %s", nvml.ErrorString(ret))
+	}
+
+	index, ret := dev.GetIndex()
+	if ret != nvml.SUCCESS {
+		return nil, fmt.Errorf("error getting index: %s", nvml.ErrorString(ret))
+	}
+
+	memInfo, ret := dev.GetMemoryInfo()
+	if ret != nvml.SUCCESS {
+		return nil, fmt.Errorf("error getting memory info: %s", nvml.ErrorString(ret))
+	}
+
+	return &Device{
+		Device:    dev,
+		SMVersion: smVersion,
+		UUID:      uuid,
+		CoreCount: cores,
+		Index:     index,
+		Memory:    memInfo.Total,
+	}, nil
+}
+
+// DeviceCache is a cache of GPU devices, with some methods to easily access devices by UUID or index
+type DeviceCache interface {
+	// GetDeviceByUUID returns a device by its UUID
+	GetDeviceByUUID(uuid string) (*Device, error)
+	// GetDeviceByIndex returns a device by its index
+	GetDeviceByIndex(index int) (*Device, error)
+	// DeviceCount returns the number of devices in the cache
+	DeviceCount() int
+	// GetSMVersionSet returns a set of all SM versions in the cache
+	GetSMVersionSet() map[uint32]struct{}
+	// GetAllDevices returns all devices in the cache
+	GetAllDevices() []*Device
+	// GetCores returns the number of cores for a device with a given UUID. Returns an error if the device is not found.
+	GetCores(uuid string) (uint64, error)
+}
+
+// deviceCache is an implementation of DeviceCache
+type deviceCache struct {
+	allDevices   []*Device
+	uuidToDevice map[string]*Device
+	smVersionSet map[uint32]struct{}
+}
+
+// NewDeviceCache creates a new DeviceCache
+func NewDeviceCache() (DeviceCache, error) {
+	lib, err := GetNvmlLib()
+	if err != nil {
+		return nil, err
+	}
+
+	return NewDeviceCacheWithOptions(lib)
+}
+
+// NewDeviceCacheWithOptions creates a new DeviceCache with an already initialized NVML library
+func NewDeviceCacheWithOptions(nvmlLib nvml.Interface) (DeviceCache, error) {
+	cache := &deviceCache{
+		uuidToDevice: make(map[string]*Device),
+		smVersionSet: make(map[uint32]struct{}),
+	}
+
+	count, ret := nvmlLib.DeviceGetCount()
+	if ret != nvml.SUCCESS {
+		return nil, fmt.Errorf("error getting device count: %s", nvml.ErrorString(ret))
+	}
+
+	for i := 0; i < count; i++ {
+		dev, ret := nvmlLib.DeviceGetHandleByIndex(i)
+		if ret != nvml.SUCCESS {
+			log.Warnf("error getting device by index %d: %s", i, nvml.ErrorString(ret))
+			continue
+		}
+
+		device, err := NewDevice(dev)
+		if err != nil {
+			log.Warnf("error creating device index %d: %s", i, err)
+			continue
+		}
+
+		cache.uuidToDevice[device.UUID] = device
+		cache.allDevices = append(cache.allDevices, device)
+		cache.smVersionSet[device.SMVersion] = struct{}{}
+	}
+
+	return cache, nil
+}
+
+// GetDeviceByUUID returns a device by its UUID
+func (c *deviceCache) GetDeviceByUUID(uuid string) (*Device, error) {
+	device, ok := c.uuidToDevice[uuid]
+	if !ok {
+		return nil, fmt.Errorf("device with uuid %s not found", uuid)
+	}
+	return device, nil
+}
```

Comment:
maybe keep the same interface as map get? return device, bool and not device, error ?